rust/lance-encoding/src/encodings/physical/dictionary.rs

```diff
@@ -0,0 +1,410 @@
+    encoder::{ArrayEncoder, EncodedArray},
+    format::pb,
+    EncodingsIo,
+};
+
+use crate::decoder::LogicalPageDecoder;
+use crate::encodings::logical::primitive::PrimitiveFieldDecoder;
+
+use arrow_schema::DataType;
+use bytes::BytesMut;
+use lance_core::Result;
+use std::collections::HashMap;
+
+use crate::encodings::utils::new_primitive_array;
+use arrow_array::cast::AsArray;
+
+#[derive(Debug)]
+pub struct DictionaryPageScheduler {
+    indices_scheduler: Arc<dyn PageScheduler>,
+    items_scheduler: Arc<dyn PageScheduler>,
+    num_dictionary_items: u32,
+}
+
+impl DictionaryPageScheduler {
+    pub fn new(
+        indices_scheduler: Arc<dyn PageScheduler>,
+        items_scheduler: Arc<dyn PageScheduler>,
+        num_dictionary_items: u32,
+    ) -> Self {
+        Self {
+            indices_scheduler,
+            items_scheduler,
+            num_dictionary_items,
+        }
+    }
+}
+
+impl PageScheduler for DictionaryPageScheduler {
+    fn schedule_ranges(
+        &self,
+        ranges: &[std::ops::Range<u64>],
+        scheduler: &Arc<dyn EncodingsIo>,
+        top_level_row: u64,
+    ) -> BoxFuture<'static, Result<Box<dyn PrimitivePageDecoder>>> {
+        // We want to decode indices and items
+        // e.g. indices [0, 1, 2, 0, 1, 0]
+        // items (dictionary) ["abcd", "hello", "apple"]
+        // This will map to ["abcd", "hello", "apple", "abcd", "hello", "abcd"]
+        // We decode all the items during scheduling itself
+        // These are used to rebuild the string later
+
+        // Schedule indices for decoding
+        let indices_page_decoder =
+            self.indices_scheduler
+                .schedule_ranges(ranges, scheduler, top_level_row);
+
+        // Schedule items for decoding
+        let items_range = 0..(self.num_dictionary_items as u64);
+        let items_page_decoder = self.items_scheduler.schedule_ranges(
+            std::slice::from_ref(&items_range),
+            scheduler,
+            top_level_row,
+        );
+
+        let copy_size = self.num_dictionary_items as u64;
+
+        tokio::spawn(async move {
+            let items_decoder: Arc<dyn PrimitivePageDecoder> = Arc::from(items_page_decoder.await?);
+
+            let mut primitive_wrapper = PrimitiveFieldDecoder::new_from_data(
+                items_decoder.clone(),
+                DataType::Utf8,
+                copy_size,
+            );
+
+            // Decode all items
+            let drained_task = primitive_wrapper.drain(copy_size)?;
+            let items_decode_task = drained_task.task;
+            let mut decoded_dict = items_decode_task.decode()?;
+
+            // For the case where the dictionary is encoded as a single null element
+            // Reset it to be an empty string array
+            if decoded_dict.len() == 1 && decoded_dict.is_null(0) {
+                decoded_dict = Arc::new(StringArray::from(Vec::<Option<String>>::new()));
+            }
```

Comment:
What's the motivation for resetting this?  This array will never be accessed in this case so what harm is there in just leaving it as it is?